experimental/parser/lexer.go

```diff
@@ -0,0 +1,675 @@
+	}
+	var start, end token.Token
+	l.All()(func(tok token.Token) bool {
+		if tok.IsSynthetic() {
+			return false
+		}
+
+		switch tok.Kind() {
+		case token.Space, token.Comment:
+			break
+		case token.String:
+			if start.Nil() {
+				start = tok
+			} else {
+				end = tok
+			}
+		default:
+			if !start.Nil() && !end.Nil() {
+				catStrings(start, end)
+			}
+			start = token.Nil
+			end = token.Nil
+		}
+		return true
+	})
+	if !start.Nil() && !end.Nil() {
+		catStrings(start, end)
+	}
+}
+
+func (l *lexer) Push(length int, kind token.Kind) token.Token {
+	l.count++
+	return l.Stream.Push(length, kind)
+}
+
+// LexString lexes a number starting at the current cursor.
+func (l *lexer) LexNumber() token.Token {
+	start := l.cursor
+	// Accept all digits, legalize later.
+	// Consume the largest prefix that satisfies the rules at
+	// https://protobuf.com/docs/language-spec#numeric-literals
+more:
+	r := l.Peek()
+	if r == 'e' || r == 'E' {
+		_ = l.Pop()
+		r = l.Peek()
+		if r == '+' || r == '-' {
+			_ = l.Pop()
+		}
+
+		goto more
+	}
+	if r == '.' || unicode.IsDigit(r) || unicode.IsLetter(r) ||
+		// We consume _ because 0_n is not valid in any context, so we
+		// can offer _ digit separators as an extension.
+		r == '_' {
+		_ = l.Pop()
+		goto more
+	}
+
+	// Create the token, even if this is an invalid number. This will help
+	// the parser pick up bad numbers into number literals.
+	digits := l.Text()[start:l.cursor]
+	tok := l.Push(len(digits), token.Number)
+
+	// Delete all _s from digits and normalize to lowercase.
+	digits = strings.ToLower(strings.ReplaceAll(digits, "_", ""))
+
+	// Now, let's see if this is actually a valid number that needs a diagnostic.
+	// First, try to reify it as a uint64.
+	switch {
+	case strings.HasPrefix(digits, "0x"):
+		value, err := strconv.ParseUint(strings.TrimPrefix(digits, "0x"), 16, 64)
+		if err == nil {
+			token.SetValue(tok, value)
+			return tok
+		}
+
+		// Emit a diagnostic. Which diagnostic we emit depends on the error Go
+		// gave us. NB: all ParseUint errors are *strconv.NumError, as promised
+		// by the documentation.
+		if err.(*strconv.NumError).Err == strconv.ErrRange {
+			l.Error(ErrIntegerOverflow{Token: tok})
+		} else {
+			l.Error(ErrInvalidNumber{Token: tok})
+		}
+	case strings.HasPrefix(digits, "0") && strings.IndexFunc(digits, func(r rune) bool { return r < '0' || r > '7' }) == -1:
+		// Annoyingly, 0777 is octal, but 0888 is not, so we have to handle this case specially.
+		fallthrough
+	case strings.HasPrefix(digits, "0o"): // Rust/Python-style octal ints are an extension.
+		value, err := strconv.ParseUint(strings.TrimPrefix(digits, "0o"), 8, 64)
+		if err == nil {
+			token.SetValue(tok, value)
+			return tok
+		}
+
+		if err.(*strconv.NumError).Err == strconv.ErrRange {
+			l.Error(ErrIntegerOverflow{Token: tok})
+		} else {
+			l.Error(ErrInvalidNumber{Token: tok})
+		}
+	case strings.HasPrefix(digits, "0b"): // Binary ints are an extension.
```

Comment:
Same remark as for supporting `_`: we can parse this for subsequent diagnostics (since we understand their intent, or at least think we do), but we should emit a diagnostic that protobuf does not allow this flavor of octal or binary numeric int literals.

_Actually_ supporting this syntax is a change to the language, which we should discuss separately.

Reply:
Same here, these have diagnostics now.